src/maxwell/maxwell_constraints.c

```diff
@@ -302,48 +306,175 @@ double *maxwell_yparity(evectmatrix X, maxwell_data *d)
      mpi_allreduce(yp_scratch, yparity, X.p,
 		   double, MPI_DOUBLE, MPI_SUM, mpb_comm);
      mpi_allreduce(norm_scratch, yp_scratch, X.p,
-		   double, MPI_DOUBLE, MPI_SUM, mpb_comm);     
+		   double, MPI_DOUBLE, MPI_SUM, mpb_comm);
      for (b = 0; b < X.p; ++b)
          yparity[b] /= yp_scratch[b];
      free(yp_scratch);
      free(norm_scratch);
-     
+
      return yparity;
 }
 
+/* Similar to the zparity functions above, but for the x -> -x mirror flip. */
+
+/* Project X to its even or odd component, so that we can solve
+   for only one parity of states (the projection operator, like the
+   mirror flip operator, commutes with the Maxwell operator, so this
+   projection should not slow convergence).  */
+void maxwell_xparity_constraint(evectmatrix X, void *data)
+{
+#ifdef HAVE_MPI
+     CHECK(NULL, "maxwell_xparity_constraint not available when using MPI.");
+#endif
+     maxwell_data *d = (maxwell_data *) data;
+     int i, j, k, b, nx, ny, nz;
+     int xparity = ((d->parity & EVEN_X_PARITY) ? +1 :
+		    ((d->parity & ODD_X_PARITY) ? -1 : 0));
+
+     if (xparity == 0)
+	  return;
+
+     CHECK(d, "null maxwell data pointer!");
+     CHECK(X.c == 2, "fields don't have 2 components!");
+
+     nx = d->nx;
+     ny = d->ny;
+     nz = d->nz;
+
+     for (i = 0; 2*i < nx; ++i) {
+       for (j = 0; j <= ny; ++j) {
+         int ij = i * ny + j;
+         int ij2 = i * ny + (j > 0 ? ny - j : 0);
+         for (k = 0; k < nz; ++k) {
+           int ijk = ij * nz + k;
+           int ijk2 = ij2 * nz + k;
+           for (b = 0; b < X.p; ++b) {
+             scalar u,v, u2,v2;
+             u = X.data[(ijk * 2) * X.p + b];
+             v = X.data[(ijk * 2 + 1) * X.p + b];
+             u2 = X.data[(ijk2 * 2) * X.p + b];
+             v2 = X.data[(ijk2 * 2 + 1) * X.p + b];
+             ASSIGN_SCALAR(X.data[(ijk * 2) * X.p + b],
+                           0.5*(SCALAR_RE(u) - xparity*SCALAR_RE(u2)),
+                           0.5*(SCALAR_IM(u) - xparity*SCALAR_IM(u2)));
+             ASSIGN_SCALAR(X.data[(ijk * 2 + 1) * X.p + b],
+                           0.5*(SCALAR_RE(v) + xparity*SCALAR_RE(v2)),
+                           0.5*(SCALAR_IM(v) + xparity*SCALAR_IM(v2)));
+             ASSIGN_SCALAR(X.data[(ijk2 * 2) * X.p + b],
+                           0.5*(SCALAR_RE(u2) - xparity*SCALAR_RE(u)),
+                           0.5*(SCALAR_IM(u2) - xparity*SCALAR_IM(u)));
+             ASSIGN_SCALAR(X.data[(ijk2 * 2 + 1) * X.p + b],
+                           0.5*(SCALAR_RE(v2) + xparity*SCALAR_RE(v)),
+                           0.5*(SCALAR_IM(v2) + xparity*SCALAR_IM(v)));
```

Comment:
This isn't right.  It's tricky, because we are mirror flipping a vector field, but the vector field is stored in a funny "transverse" coordinate system in order to impose a key mathematical constraint. 

The "transverse" coordinate system is specially chosen to have nice properties for mirror flipping in the y and z direction: https://github.com/stevengj/mpb/blob/62495d4f9b3402b31383f0cc255cb5a95d033f7e/src/maxwell/maxwell.c#L398-L399

but it may not be so nice for flipping in x.